src/Paramore.Brighter/Transforms/Storage/IAmAStorageProviderAsync.cs

```diff
@@ -0,0 +1,69 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE. */
+
+#endregion
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Paramore.Brighter.Transforms.Storage
+{
+    /// <summary>
+    /// Provides storage for a message body that is too large to transmit over message-oriented middleware
+    /// The stored value is the 'luggage' and the id of the item is the 'claim check'
+    /// </summary>
+    public interface IAmAStorageProviderAsync
+    {
+        /// <summary>
+        /// Delete the luggage identified by the claim check
+        /// Used to clean up after luggage is retrieved
+        /// </summary>
+        /// <param name="claimCheck">The claim check for the luggage</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        Task DeleteAsync(string claimCheck, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Downloads the luggage associated with the claim check
+        /// </summary>
+        /// <param name="claimCheck">The claim check for the luggage</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>The luggage as a stream</returns>
+        Task<Stream> DownloadAsync(string claimCheck, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Do we have luggage for this claim check - in case of error or deletion
+        /// </summary>
+        /// <param name="claimCheck"></param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        Task<bool> HasClaimAsync(string claimCheck, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Puts luggage into the store and provides a claim check for that luggage
+        /// </summary>
+        /// <param name="stream">A stream representing the luggage to check</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>A claim check for the luggage stored</returns>
+        Task<string> UploadAsync(Stream stream, CancellationToken cancellationToken);
```

Comment:
Not a big one, do we want to use Upload and Download or Store and Retrieve, I suppose in the distributed world we are always going to be "uploading" and there wouldn't be a reason to put it on local disk ect. so probably not a problem,